node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+        descriptor_bytes: &impl AsRef<[u8]>,
+    ) -> Self {
+        Descriptor {
+            proto_root,
+            dependencies,
+            descriptor_proto: prost_types::FileDescriptorSet::decode(descriptor_bytes.as_ref())
+                .unwrap(),
+        }
+    }
+
+    /// Loads the descriptor to the pool, if not already loaded.
+    pub fn load(&self, pool: &mut prost_reflect::DescriptorPool) -> anyhow::Result<()> {
+        if self
+            .descriptor_proto
+            .file
+            .iter()
+            .all(|f| pool.get_file_by_name(f.name()).is_some())
+        {
+            return Ok(());
+        }
+        for d in &self.dependencies {
+            d.load(pool)?;
+        }
+        pool.add_file_descriptor_set(self.descriptor_proto.clone())?;
+        Ok(())
+    }
+
+    /// Loads the descriptor to the global pool and returns a copy of the global pool.
+    pub fn load_global(&self) -> prost_reflect::DescriptorPool {
+        /// Global descriptor pool.
+        static POOL: Lazy<Mutex<prost_reflect::DescriptorPool>> = Lazy::new(Mutex::default);
+        let pool = &mut POOL.lock().unwrap();
+        self.load(pool).unwrap();
+        pool.clone()
+    }
+}
+
+/// Code generation config. Use it in build scripts.
+pub struct Config {
+    /// Input directory relative to $CARGO_MANIFEST_DIR with the proto files to be compiled.
+    pub input_root: InputPath,
+    /// Implicit prefix that should be prepended to proto paths of the proto files in the input directory.
+    pub proto_root: ProtoPath,
+    /// Descriptors of the dependencies and the rust absolute paths under which they will be available from the generated code.
+    pub dependencies: Vec<(RustName, &'static Descriptor)>,
+    /// Rust absolute path under which the protobuf crate will be available from the generated
+    /// code.
+    pub protobuf_crate: RustName,
+}
+
+impl Config {
+    /// Location of the protobuf_build crate, visible from the generated code.
+    fn this_crate(&self) -> RustName {
+        self.protobuf_crate.clone().join("build")
+    }
+
+    /// Generates implementation of `prost_reflect::ReflectMessage` for a rust type generated
+    /// from a message of the given `proto_name`.
+    fn reflect_impl(&self, proto_name: ProtoName) -> String {
+        let rust_name = proto_name
+            .relative_to(&self.proto_root.to_name())
+            .unwrap()
+            .to_rust_type()
+            .to_string();
+        let proto_name = proto_name.to_string();
+        let this = self.this_crate().to_string();
+        format!("impl {this}::prost_reflect::ReflectMessage for {rust_name} {{\
+            fn descriptor(&self) -> {this}::prost_reflect::MessageDescriptor {{\
+                static INIT : {this}::Lazy<{this}::prost_reflect::MessageDescriptor> = {this}::Lazy::new(|| {{\
+                    DESCRIPTOR.load_global().get_message_by_name({proto_name:?}).unwrap()\
+                }});\
+                INIT.clone()\
```

Comment:
- As with the descriptor definition, ideally we'd want to move as much code from here to the crate itself as possible. Again, this'll probably require using a macro like `derive_reflect_message!($message:ty, $descriptor:ident)`.
- Stupid question: since `DESCRIPTOR` is a static already, does it make sense to cache separate `MessageDescriptor`s here instead of just returning a fresh clone each time? Is it significantly more efficient?

Reply:
load_global() is inefficient because it checks whether all the files of the descriptor have been already loaded. Plus, cloning prost_reflect::DescriptorPool in load_global() is inefficient because it is poorly written.